streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -227,21 +253,72 @@ public final TopologyBuilder addSource(String name, Deserializer keyDeserializer
 
+        if (topicPattern != null) {
+            throw new TopologyBuilderException("Subscription to topics, partitions and pattern are mutually exclusive");
+        }
+
         for (String topic : topics) {
             if (sourceTopicNames.contains(topic))
                 throw new TopologyBuilderException("Topic " + topic + " has already been registered by another source.");
 
             sourceTopicNames.add(topic);
         }
 
-        nodeFactories.put(name, new SourceNodeFactory(name, topics, keyDeserializer, valDeserializer));
+        nodeFactories.put(name, new SourceNodeFactory(name, topics, null, keyDeserializer, valDeserializer));
         nodeToSourceTopics.put(name, topics.clone());
         nodeGrouper.add(name);
 
         return this;
     }
 
     /**
+     * Add a new source that consumes from topics matching the given pattern
+     * and forwards the records to child processor and/or sink nodes.
+     * The source will use the specified key and value deserializers.
+     *
+     * @param name the unique name of the source used to reference this node when
+     * {@link #addProcessor(String, ProcessorSupplier, String...) adding processor children}.
+     * @param keyDeserializer the {@link Deserializer key deserializer} used when consuming records; may be null if the source
+     * should use the {@link org.apache.kafka.streams.StreamsConfig#KEY_SERDE_CLASS_CONFIG default key deserializer} specified in the
+     * {@link org.apache.kafka.streams.StreamsConfig stream configuration}
+     * @param valDeserializer the {@link Deserializer value deserializer} used when consuming records; may be null if the source
+     * should use the {@link org.apache.kafka.streams.StreamsConfig#VALUE_SERDE_CLASS_CONFIG default value deserializer} specified in the
+     * {@link org.apache.kafka.streams.StreamsConfig stream configuration}
+     * @param topicPattern regular expression pattern to match Kafka topics that this source is to consume
+     * @return this builder instance so methods can be chained together; never null
+     * @throws TopologyBuilderException if processor is already added or if topics have already been registered by name
+     */
+
+    public final TopologyBuilder addSource(String name, Deserializer keyDeserializer, Deserializer valDeserializer, Pattern topicPattern) {
+
+        if (topicPattern == null) {
+            throw new TopologyBuilderException("Pattern can't be null");
+        }
+
+        if (nodeFactories.containsKey(name)) {
+            throw new TopologyBuilderException("Processor " + name + " is already added.");
+        }
+
+        if (!sourceTopicNames.isEmpty()) {
+            throw new TopologyBuilderException("Subscription to topics, partitions and pattern are mutually exclusive");
+        }
+
+        if (this.topicPattern != null) {
+            throw new TopologyBuilderException("Pattern source already defined");
```

Comment:
nit: "Only one source node with regex subscription pattern can be defined in the topology, and there is an existing pattern: {}".